### PR TITLE
`solvers`: ParaSwap API

### PR DESCRIPTION
**Pull request by cowfee**
Progress on https://github.com/cowprotocol/services/issues/882.

Add ParaSwap API to the `solvers` crate.

### Test Plan

<details>
<summary>Ran a manual test:</summary>

```rs
#[cfg(test)]
mod tests {
    use {
        super::*,
        crate::domain::{auction, order},
    };

    #[tokio::test]
    async fn manual() {
        shared::tracing::initialize_reentrant("solvers=trace");

        let dex = ParaSwap {
            client: reqwest::Client::new(),
            config: Config {
                endpoint: DEFAULT_URL.parse().unwrap(),
                exclude_dexs: Default::default(),
                address: "0xE0B3700e0aadcb18ed8d4BFF648Bc99896a18ad1"
                    .parse()
                    .unwrap(),
                partner: Some("Test".to_owned()),
            },
        };

        let swap = dex
            .swap(
                &dex::Order {
                    sell: eth::TokenAddress(
                        "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"
                            .parse()
                            .unwrap(),
                    ),
                    buy: eth::TokenAddress(
                        "0xE41d2489571d322189246DaFA5ebDe1F4699F498"
                            .parse()
                            .unwrap(),
                    ),
                    side: order::Side::Sell,
                    amount: dex::Amount::new(1_000_000_000_000_000_000_u128.into()),
                },
                &dex::Slippage::one_percent(),
                &auction::Tokens(
                    [
                        (
                            eth::TokenAddress(
                                "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"
                                    .parse()
                                    .unwrap(),
                            ),
                            auction::Token {
                                decimals: Some(18),
                                symbol: None,
                                reference_price: None,
                                available_balance: Default::default(),
                                trusted: false,
                            },
                        ),
                        (
                            eth::TokenAddress(
                                "0xE41d2489571d322189246DaFA5ebDe1F4699F498"
                                    .parse()
                                    .unwrap(),
                            ),
                            auction::Token {
                                decimals: Some(18),
                                symbol: None,
                                reference_price: None,
                                available_balance: Default::default(),
                                trusted: false,
                            },
                        ),
                    ]
                    .into_iter()
                    .collect(),
                ),
            )
            .await
            .unwrap();

        panic!("{swap:#?}");
    }
}
```
</details>

### CODE REVIEW

**Review by vkgnosis — APPROVED** [review]

File: crates/solvers/src/infra/dex/mod.rs
Comment (MartinquaXD) [inline crates/solvers/src/infra/dex/mod.rs:30]: I think only paraswap actually needs the additional metadata of the auction to get the decimals. All the other implementations could continue to just take the gas price.

File: crates/solvers/src/infra/dex/balancer/mod.rs
Comment (vkgnosis) [reply crates/solvers/src/infra/dex/balancer/mod.rs:47]: I think it's fine to move code around if it makes more sense in the end but it would be nice to have refactors unrelated to the main change in a separate commit. Not a requirement, just nice.

File: crates/solvers/src/domain/dex/slippage.rs
Comment (cowfee) [reply crates/solvers/src/domain/dex/slippage.rs:115]: If you change an interface of a module in `driver`, and refactor the code without changing behavior, you will not need to change any tests.

File: crates/solvers/src/infra/dex/paraswap/mod.rs
Comment (cowfee) [reply crates/solvers/src/infra/dex/paraswap/mod.rs:9]: It will be used soon :smile:

File: crates/solvers/src/infra/dex/paraswap/dto.rs
Comment (cowfee) [reply crates/solvers/src/infra/dex/paraswap/dto.rs:157]: The raw `serde_json::Value` inside `PriceRoute` needs to be passed into the `/transaction` endpoint.

In the [legacy code](https://github.com/cowprotocol/services/blob/192399adbc81d9a3821f71caf6cca02c0f23d752/crates/shared/src/paraswap_api.rs#L206-L246) there is an explicit implementation of `Deserialize`. I think this approach is slightly simpler, but YMMV. I don't think it matters that much. I can change this if you want.

File: crates/solvers/src/infra/dex/mod.rs
Comment (cowfee) [reply crates/solvers/src/infra/dex/mod.rs:30]: Why would they? The `Auction` is available to the DEX `swap` process. I'm keeping all of the interfaces consistent.

File: crates/solvers/src/infra/dex/balancer/mod.rs
Comment (cowfee) [reply crates/solvers/src/infra/dex/balancer/mod.rs:47]: I find it easier to read if the `pub` methods come first. I can undo these changes if you want.

File: crates/solvers/src/domain/dex/slippage.rs
Comment (cowfee) [reply crates/solvers/src/domain/dex/slippage.rs:123]: > We pass it in like that because it's easier at the place the function is used.

No, we pass it in like that because we're building coherent abstractions. `slippage::Prices` is not a coherent abstraction.

File: crates/solvers/src/boundary/legacy.rs
Comment (cowfee) [reply crates/solvers/src/boundary/legacy.rs:244]: https://github.com/cowprotocol/services/pull/1513/commits/3d9a6ca963c6c890eb40393fd0d228be4bba5886

File: crates/solvers/src/domain/auction.rs
Comment (cowfee) [reply crates/solvers/src/domain/auction.rs]: https://github.com/cowprotocol/services/pull/1513/commits/3d9a6ca963c6c890eb40393fd0d228be4bba5886

File: crates/solvers/src/infra/dex/paraswap/mod.rs
Comment (cowfee) [reply crates/solvers/src/infra/dex/paraswap/mod.rs]: https://github.com/cowprotocol/services/pull/1513/commits/8f259bd776481d6c556ade9aa6a7d69245d2661c

File: crates/solvers/src/domain/dex/slippage.rs
Comment (MartinquaXD) [reply crates/solvers/src/domain/dex/slippage.rs:123]: > ...but now that we see that decimals are needed as well as the reference prices and the gas price from the auction, we have a pretty strong indicator that we should just use `Auction`.

Where in `Limits` do you need the decimals and gas price? IMO computing slippage is very far detached from the concept of an `Auction`. Only the reference prices happen to be in an auction but from the perspective of the `Limits` code it doesn't matter where the reference price comes from.

File: crates/solvers/src/infra/dex/mod.rs
Comment (MartinquaXD) [reply crates/solvers/src/infra/dex/mod.rs:30]: IMO it makes sense to only pass around what you actually need in a function.
To me it's a deficiency of the paraswap API that you have to specify the decimals with your request which then leaks into the remaining dex implementations.
I understood the `infra` crate to exist just so we can interface with external APIs and for those the concept of an `Auction` is meaningless.

File: crates/solvers/src/infra/dex/paraswap/dto.rs
Comment (MartinquaXD) [reply crates/solvers/src/infra/dex/paraswap/dto.rs:173]: Cloning unnecessarily probably isn't a big deal. 👌

File: crates/solvers/src/infra/dex/balancer/mod.rs
Comment (MartinquaXD) [reply crates/solvers/src/infra/dex/balancer/mod.rs:47]: You don't have to change the PR but then I would like it if the team could agree on some basic rules that would minimize the need to move around code to preserve the git history as much as possible.

File: crates/solvers/src/domain/order.rs
Comment (nlordell) [inline crates/solvers/src/domain/order.rs:105]: I think that is a better name.

File: crates/solvers/src/domain/dex/slippage.rs
Comment (nlordell) [reply crates/solvers/src/domain/dex/slippage.rs:123]: My rationale for creating a `dex::slippage::Prices` utility type was specifically because slippage calculation does not *want* auction prices, but token exchange rates to Ether, so it is conceptually different from auction reference prices.

Furthermore, it felt icky from an abstraction point of view to pass in a complete auction for slippage computation when all you need from an auction are its reference prices for computing Ether exchange rates. Auctions include orders, liquidity, deadline, token metadata, etc. which aren't really needed for slippage computation (and in the function signature is unclear . This manifests itself in the various `Default::default()` calls when setting up the unit test 👇. What would be slightly cleaner here IMO would be to pass in `Auction::prices` and not require the utility type anymore.

The unit test was included here as it testing this from the `Solver` boundary is quite difficult. I agree that having refactors require annoying changes to unit tests is sub-optimal.

File: crates/solvers/src/domain/solver/dex/solver.rs
Comment (nlordell) [inline crates/solvers/src/domain/solver/dex/solver.rs:1]: `mv dex/solver.rs dex/mod.rs` makes sense. I don't know what I was thinking before 😛.

File: crates/solvers/src/infra/dex/paraswap/mod.rs
Comment (nlordell) [reply crates/solvers/src/infra/dex/paraswap/mod.rs:64]: No, this is incorrect. The spender is not the `to` address but there should be an `allowance_target` or something (in fact, giving an allowance to the ParaSwap router will allow others to steal Settlement contract funds)

File: crates/solvers/src/infra/dex/zeroex/dto.rs
Comment (nlordell) [inline crates/solvers/src/infra/dex/zeroex/dto.rs:80]: This was mostly to keep a `defaults` around with the default query fields. It should probably be split into two types to be more correct.

File: crates/solvers/src/infra/dex/paraswap/mod.rs
Comment (nlordell) [reply crates/solvers/src/infra/dex/paraswap/mod.rs:64]: Just checked and it should be the `tokenTransferProxy` field from the price route response.

File: crates/solvers/src/infra/dex/mod.rs
Comment (nlordell) [inline crates/solvers/src/infra/dex/mod.rs:25]: Not passing in an `Auction` was also an intentional thing I had from before. The rationale here is you end up passing in data that may be "inconsistent", and it is unclear from the caller's perspective what the expectations are:
- What happens if there is no order in the auction that is equivalent to the order parameter?
- What do you do with all the liquidity?
- Does the callee check the deadline?
- etc.

Essentially, it makes the function operate on a lot more data that it actually cares about and creates some ambiguity on what the function actually does with the additional data.

File: crates/solvers/src/infra/dex/mod.rs
Comment (nlordell) [reply crates/solvers/src/infra/dex/mod.rs:30]: I had a similar comment here: https://github.com/cowprotocol/services/pull/1513/files#r1198763872

Maybe, instead of `dex::slippage::Prices` you need a `dex::Auction` domain type that models the additional auction data that DEX-based solvers care about (so the reference prices and token metadata)?

**Review by nlordell — COMMENTED** [review]
PR looks good in general, left some comments.

File: crates/solvers/src/domain/dex/slippage.rs
Comment (nlordell) [reply crates/solvers/src/domain/dex/slippage.rs:115]: In general, I also think it is annoying to have to modify unit tests for refactors (it causes unnecessary additional barriers to refactoring that may dissuade people from doing the refactor they want for the additional work).

On the other hand, in this particular case I think its fine - it just caused changes to unit tests in the same module that was refactored (so, it can

In theory this _can_ be tested at the `solvers` boundary, it would be a lot harder though (and arguably less "correct", as integration tests assertions may require assumptions about the details of the implementation).

File: crates/solvers/src/domain/auction.rs
Comment (nlordell) [reply crates/solvers/src/domain/auction.rs]: ParaSwap API is annoying in this regard - as it expects the client to send token metadata along with the request. :/

**Comment by cowfee** [comment]
@MartinquaXD @nlordell Thanks for the insightful reviews! I was wrong, passing the `Auction` is too much and it creates more questions than it resolves :smile:

Added an `auction::Tokens` type and passed that instead of the full auction. https://github.com/cowprotocol/services/pull/1513/commits/aeb6e3bc819f272f013c007647b143c54de351a4

File: crates/solvers/src/domain/auction.rs
Comment (nlordell) [inline crates/solvers/src/domain/auction.rs:20]: Nice, I think this is more coherent than the existing `dex::slippage::Prices` and the previously proposed passing in `auction::Auction`. Improvement to the `domain`!

File: crates/solvers/src/domain/auction.rs
Comment (nlordell) [inline crates/solvers/src/domain/auction.rs:67]: nit: I would undo this change. `default` gas price seems to be neede for unit test convenience, but I don't think it conceptually makes sense (is the default 0? this won't work with most chain, not even test ones with default parameters).

File: crates/solvers/src/domain/dex/slippage.rs
Comment (nlordell) [inline crates/solvers/src/domain/dex/slippage.rs:103]: This code is pre-nick-spell-check-reform.

File: crates/solvers/src/infra/dex/balancer/mod.rs
Comment (nlordell) [reply crates/solvers/src/infra/dex/balancer/mod.rs:47]: One advantage to "allowing" refactors and moving code around within a PR is that it makes it more convenient to refactor the core logic of the services (and removes additional barriers to keeping the abstractions "up to date", no need to go through the hassle for creating additional PRs for example).

As @MartinquaXD suggested, maybe we can come up with some "standard practices" for code ordering to make these less common?

File: crates/solvers/src/infra/dex/mod.rs
Comment (nlordell) [inline crates/solvers/src/infra/dex/mod.rs:32]: Interesting, ParaSwap is the only API that doesn't allow user-specified gas parameters for its route finding.

Not comment on the code per se, just something interesting I noticed (so, if as a user I want to submit a "high gas" swap, PAraSwap may pick a sub-optimal route).

Comment (cowfee) [reply crates/solvers/src/infra/dex/mod.rs:32]: Yeah, weird...

File: crates/solvers/src/infra/dex/paraswap/dto.rs
Comment (nlordell) [inline crates/solvers/src/infra/dex/paraswap/dto.rs:185]: The `tokenProxyAddress` is also needed from the `priceRoute` response to get the approval target.

File: crates/solvers/src/infra/dex/balancer/mod.rs
Comment (cowfee) [reply crates/solvers/src/infra/dex/balancer/mod.rs:47]: I just play it by ear for the most part. I want everybody to feel encouraged to constantly refactor the code, otherwise the code will rot as you add new features. This inevitably hurts git blame but it's worth it.

**Review by nlordell — APPROVED** [review]
Nice, only issue I still see is the approval target is incorrect. Approving **assuming that it will be fixed before merging**.

Also, small nit on `GasPrice: Default` implementation.

File: crates/solvers/src/infra/dex/mod.rs
Comment (cowfee) [reply crates/solvers/src/infra/dex/mod.rs]: https://github.com/cowprotocol/services/pull/1513/commits/aeb6e3bc819f272f013c007647b143c54de351a4

File: crates/solvers/src/infra/dex/mod.rs
Comment (nlordell) [reply crates/solvers/src/infra/dex/mod.rs:32]: It appears to only allow specifying one when building the transaction and not when finding the price route (at which point, it is kind of useless anyway).

File: crates/solvers/src/domain/solver/dex/solver.rs
Comment (nlordell) [reply crates/solvers/src/domain/solver/dex/solver.rs:1]: The hint was in `solver::dex::solver` module path 😛

**Review by MartinquaXD — APPROVED** [review]
Also assuming the paraswap allowance spender address gets fixed.

File: crates/solvers/src/infra/dex/paraswap/mod.rs
Comment (cowfee) [reply crates/solvers/src/infra/dex/paraswap/mod.rs]: https://github.com/cowprotocol/services/pull/1513/commits/00768912d9f05ec356f8912f2add1c9dd63ebe9a

File: crates/solvers/src/domain/order.rs
Comment (cowfee) [reply crates/solvers/src/domain/order.rs]: https://github.com/cowprotocol/services/pull/1513/commits/346b2300c078e82d26f819c0182d359bd3c3d3e3

File: crates/solvers/src/infra/dex/mod.rs
Comment (cowfee) [reply crates/solvers/src/infra/dex/mod.rs]: https://github.com/cowprotocol/services/pull/1513/commits/aeb6e3bc819f272f013c007647b143c54de351a4

File: crates/solvers/src/domain/auction.rs
Comment (cowfee) [reply crates/solvers/src/domain/auction.rs]: https://github.com/cowprotocol/services/pull/1513/commits/32cc414ac5be918d476da86ee75ebf9b4ce1a63a

**Pull request merged**
